lib/flight_scheduler/schedulers/backfilling_scheduler.rb

```diff
@@ -0,0 +1,261 @@
+
+      # We've exited the allocation loop. Any jobs left 'WaitingForScheduling'
+      # are blocked on priority.  We'll update a few of them to show that is
+      # the case.
+      #
+      # A more complicated scheduler would likely do this whilst iterating
+      # over the jobs.
+      candidates
+        .take(5)
+        .select { |job| job.reason_pending == 'WaitingForScheduling' }
+        .each { |job| job.reason_pending = 'Priority' }
+    end
+    new_allocations
+  end
+
+  def queue
+    # For the backfilling scheduler, the queue is sorted and slightly filtered
+    # view of the registry of jobs.
+    #
+    # The sorting and filtering is primarily intended for display purposes,
+    # but also using it for processing ensures that the jobs are processed in
+    # the order in which the queue output suggests.
+    #
+    # The jobs are sorted to ensure that running jobs are placed higher in the
+    # queue output than pending jobs.
+    #
+    # The jobs are filtered to ensure that array jobs without any pending
+    # tasks are not included.
+    #
+    # A more complicated scheduler may implement a priority queue and need to
+    # keep its own record of the queued jobs rather than using job registry.
+    running_jobs_first = lambda { |job| job.reason_pending.nil? ? -1 : 1 }
+
+    # NOTE: We rely on the sort being stable to ensure that earlier added jobs
+    # are considered prior to later added jobs.
+    FlightScheduler.app.job_registry.jobs
+      .reject { |j| j.job_type == 'ARRAY_JOB' && j.task_generator.finished? }
+      .reject { |j| j.terminal_state? }
+      .sort_by.with_index { |x, idx| [running_jobs_first.call(x), idx] }
+  end
+
+  # Attempt to allocate a single job.
+  #
+  # If the partition has sufficient resources available for the job, a
+  # new +Allocation+ is added to the allocations registry and returned.
+  #
+  # If an allocation is created for an `ARRAY_TASK`, it is added to the job
+  # registry and the associated task generator advanced.
+  #
+  # If there are insufficient resources available to allocate to the job,
+  # the job's `reason_pending` is updated and +nil+ is returned.
+  #
+  # NOTE: It is expected that this will not differ between schedulers.
+  def allocate_job(job, reason: 'Resources')
+    raise InvalidJobType, job if job.job_type == 'ARRAY_JOB'
+
+    # Generate an allocation for the job
+    nodes = job.partition.nodes
+    FlightScheduler::LoadBalancer.new(job: job, nodes: nodes).allocate.tap do |allocation|
+      if allocation.nil?
+        if job.job_type == 'ARRAY_TASK'
+          job.array_job.reason_pending = reason
+        else
+          job.reason_pending = reason
+        end
+        nil
+      else
+        if job.job_type == 'ARRAY_TASK'
+          FlightScheduler.app.job_registry.add(job)
+          job.array_job.task_generator.advance_next_task
+        end
+        FlightScheduler.app.allocations.add(allocation)
+      end
+    end
+  end
+
+  def candidates
```

Comment:
This looks like a pretty general method, might be worth considering extracting it at some point. It would be a convenient way to handle time limits in the `FiFoScheduler`.

Reply:
That was very much the intention.  Not a change I want to make right now though.  Too little time.